src/scales/scale.linearbase.js

```diff
@@ -50,23 +61,40 @@ function generateTicks(generationOptions, dataRange) {
   niceMin = Math.floor(rmin / spacing) * spacing;
   niceMax = Math.ceil(rmax / spacing) * spacing;
 
-  // If min, max and stepSize is set and they make an evenly spaced scale use it.
-  if (stepSize && minDefined && maxDefined) {
-    // If very close to our whole number, use it.
-    if (almostWhole((max - min) / stepSize, spacing / 1000)) {
-      niceMin = min;
-      niceMax = max;
-    }
-  }
-
-  numSpaces = (niceMax - niceMin) / spacing;
-  // If very close to our rounded value, use it.
-  if (almostEquals(numSpaces, Math.round(numSpaces), spacing / 1000)) {
-    numSpaces = Math.round(numSpaces);
+  if (minDefined && maxDefined && step && almostWhole((max - min) / step, spacing / 1000)) {
+    // Case 1: If min, max and stepSize are set and they make an evenly spaced scale use it.
+    // spacing = step;
+    // numSpaces = (max - min) / spacing;
+    numSpaces = Math.min((max - min) / spacing, maxTicks);
+    spacing = (max - min) / numSpaces;
+    niceMin = min;
+    niceMax = max;
+  } else if (countDefined) {
+    // Cases 2 & 3, we have a count specified. Handle optional user defined edges to the range.
+    const cMin = minDefined ? min : niceMin;
+    const cMax = maxDefined ? max : niceMax;
+    numSpaces = count - 1;
+    spacing = (cMax - cMin) / numSpaces;
+
+    // Sometimes these are no-ops, but it makes the code a lot clearer
+    // and when a user defined range is specified, we want the correct ticks
+    niceMin = cMin;
+    niceMax = cMax;
   } else {
-    numSpaces = Math.ceil(numSpaces);
+    // Case 4
+    numSpaces = (niceMax - niceMin) / spacing;
+
+    // If very close to our rounded value, use it.
+    if (almostEquals(numSpaces, Math.round(numSpaces), spacing / 1000)) {
+      numSpaces = Math.round(numSpaces);
+    } else {
+      numSpaces = Math.ceil(numSpaces);
+    }
   }
 
+  // The spacing will have changed in cases 1, 2, and 3 so the factor cannot be computed
+  // until this point
+  factor = Math.pow(10, isNullOrUndef(precision) ? _decimalPlaces(spacing) : precision);
   niceMin = Math.round(niceMin * factor) / factor;
   niceMax = Math.round(niceMax * factor) / factor;
```

Comment:
I thought about splitting the rest of this method out to another function since all it does is run the generation loop, but decided it wouldn't help the code size at all